legate/core/store.py

```diff
@@ -246,13 +254,17 @@ def detach_external_allocation(
         assert self.attached_alloc is not None
         detach = attachment_manager.remove_detachment(self.detach_key)
         detach.unordered = unordered  # type: ignore[union-attr]
-        attachment_manager.detach_external_allocation(
+        detach_future = attachment_manager.detach_external_allocation(
             self.attached_alloc, detach, defer
         )
         self.physical_region = None
         self.physical_region_mapped = False
         self.physical_region_refs = 0
         self.attached_alloc = None
+        if unordered:
+            self.detach_future = detach_future
+            if self.detach_future is not None:
+                self.field.add_detach_future(self.detach_future)
```

Comment:
I think you can probably just get rid of `RegionField.detach_future` at this point, and just set `self.field` directly:

```suggestion
        if detach_future is not None:
            self.field.add_detach_future(detach_future)
```